### PR TITLE
Add Chinese Postman UI

### PR DESCRIPTION
**Pull request by danduk82**
This PR adds the support of Chinese Postman feature and the relative UI section.
I have refactored `optimalroutebottombar` and extracted a lot of its logic to `valhallaroutebottombar`, a meta class PY file. Then I have written `cpbottombar` to extend the `valhallaroutebottombar`. 

NB: my editor has made a `black` update on a bunch of files, so sorry if this review is quite polluted by a lot of styling adaption. In the end I have left them because I thing the code is a bit more readable anyway. But for the review, focus on the following files:
- `gui/cpbottombar.py`
- `gui/optimalroutebottombar.py`
- `gui/valhallaroutebottombar.py`
- `core/optimalroutelayer.py`
- `valhalla/client.py`
- `valhalla/connectors.py`

### CODE REVIEW

File: kadasrouting/valhalla/client.py [hunk 1]
```diff
@@ -27,7 +27,7 @@ def __init__(self, connector=None):
     def isAvailable(self):
         return self.connector.isAvailable()
 
-    def route(self, qgspoints, profile, avoid_polygons, options):
+    def route(self, qgspoints, profile, avoid_polygons, options, patrol_polygons=[]):
```
Comment (ismailsunni) [inline kadasrouting/valhalla/client.py:30]: Using a list as a default parameter can introduce a gotcha. See https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments

File: kadasrouting/gui/valhallaroutebottombar.py [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+    """
+
+    def __init__(self, canvas, action, plugin):
+        KadasBottomBar.__init__(self, canvas, "orange")
+        self.setupUi(self)
+        self.setStyleSheet("QFrame { background-color: orange; }")
+        self.action = action
+        self.plugin = plugin
+        self.canvas = canvas
+        self.waypoints = []
```
Comment (ismailsunni) [inline kadasrouting/gui/valhallaroutebottombar.py:48]: should we put the waypoints on the optimalroutebottombar.py, since it's not in the CP dialog?

**Review by ismailsunni — COMMENTED** [review]
Thanks. I have two comments. Can you share the screenshot for the CP dialog also?

File: kadasrouting/gui/cpbottombar.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+            # No areas to avoid
+            patrolArea = None
+            patrolAreaWGS = None
+
+        # transform to WGS84 (Valhalla's requirement)
+        if patrolArea:
+            patrolAreaJson = json.loads(patrolArea.asJson())
+            patrolAreaWGS = []
+            polygon = patrolAreaJson["coordinates"][0]
+            for point in polygon:
```
Comment (volaya) [inline kadasrouting/gui/cpbottombar.py:137]: Why do you do this point-by-point transformation? self.patrolArea is a QgsGeometry, so you can directly transform it

Comment (ismailsunni) [reply kadasrouting/gui/cpbottombar.py:137]: I think it was from my PR: https://github.com/camptocamp/kadas-routing-plugin/pull/57/files#diff-5b0bb1792a944cff74b00cc906bb0ff6e7efa23a7f559e8c041e2b7efa952618R235

Honestly, I can't remember why I did it. I only remember that the type of geometry is a MultiPolygon, but Valhalla accepts Polygon type only. But I agree with Victor here.

File: kadasrouting/valhalla/client.py (hunk 1 above)
Comment (danduk82) [reply kadasrouting/valhalla/client.py:30]: I can put None, anyway the test is `if patrol_polygons` which goes both with an empty list or `None`

File: kadasrouting/gui/valhallaroutebottombar.py (hunk 2 above)
Comment (danduk82) [reply kadasrouting/gui/valhallaroutebottombar.py:48]: It is _not yet_ in the CP dialog. I'm pretty sure there is the risk that they might want it in the future...

File: kadasrouting/gui/cpbottombar.py (hunk 3 above)
Comment (danduk82) [reply kadasrouting/gui/cpbottombar.py:137]: good point, I'll change it and test if there are problems.

File: kadasrouting/valhalla/client.py (hunk 1 above)
Comment (ismailsunni) [reply kadasrouting/valhalla/client.py:30]: Yes, None is better in this case. Thanks!

File: kadasrouting/gui/cpbottombar.py (hunk 3 above)
Comment (danduk82) [reply kadasrouting/gui/cpbottombar.py:137]: from what I see, if you transform it directly then you get an error `QgsGeometry is not json serializable` while if we stick with the explicit list of points we don't have to do it later. I think I will leave it like this.

**Review by ismailsunni — APPROVED** [review]
LGTM

**Pull request merged**